### ISSUE TITLE
Avoid IRI lexicon join when Hop Inverse is directly within Hop with Field

### ISSUE THREAD

**Issue by brent-hartwig** [project-lux/lux-marklogic#304]
**Problem Description:**
On 27 Aug 24, a particular search led to a v8 engine crash.  This incident is being tracked by MarkLogic Support ticket no. 37229 and https://git.yale.edu/lux-its/ml-cluster-formation/issues/52.  This ticket is tracking an optimization found while investigating the crash.  The optimization may also take stress off of the v8 engine.

Original search criteria:

```json
{
  "_scope":"event",
  "used":{
    "containingItem":{
      "memberOf":{
        "name":"peabody"
      }
    }
  }
}
```

It is the `used.containingItem` portion of the query that is this ticket's focus.  The search pattern of `used` is Hop with Field.  The search pattern of `containingItem` is Hop Inverse.  Hop with Field needs IRIs, and uses the IRI lexicon to get them.  Yet, the Hop Inverse pattern is able to serve up IRIs, rendering Hop with Field's IRI lexicon join a waste of time when the input to Hop with Field is the output of Hop Inverse.

Here's an excerpt of the above search criteria's query, broken down into bits where the left side is the pre-optimized version and the right is the post-optimized version:

![image](https://github.com/user-attachments/assets/63d850e0-e6d2-4f1b-a172-b1e54bd7c6a9)

Both versions return the same results, whether filtered or not.

This optimization was the second idea.  The first idea was to add a type constraint on the call cts.triples.  I thought it might help as `la('member_of')` can be defined in Items and Sets.  It didn't help in terms of performance.  Unknown if it would help on the v8 front.

All versions of the full query: [query-versions-for-ticket-304.zip](https://github.com/user-attachments/files/16850961/query-versions-for-ticket-304.zip)

**Expected Behavior/Solution:**
Implement the above-described optimization.  Receive the same search results, but faster.  Unknown if this optimization scales linearly but should help all that have a Hop Inverse term as a direct child of a Hop with Field term.  

Scope includes determining if the optimization reduces the ability to reproduce the v8 engine crash, and reporting back to https://git.yale.edu/lux-its/ml-cluster-formation/issues/52.

**Requirements:**
Nothing additional to add here.

**Needed for promotion:**
If an item on the list is not needed, it should be crossed off but not removed.

~- [ ] Wireframe/Mockup - Mike~
- [ ] Committee discussions - Sarah
- [ ] Feasibility/Team discussion - Sarah
- [x] Backend requirements - Implement the optimization and prove it's good to go
- [x] Middle tier requirements - None
- [x] Frontend requirements - None
- [ ] Compels a performance test?
- [ ] Are new regression tests required for QA - Amy
- [ ] Questions
- List of questions for discussions. Answers should be documented within the issue.

**UAT/LUX Examples:**

- See example in opening description.
- Developer could also provide a list of term UI labels for both Hop with Field and Hop Inverse terms --the most thorough would be the available Hop with Field / Hop Inverse parent-child combinations by scope.

**Dependencies/Blocks:**

This issue neither blocks nor is blocked by another ticket.

**Related Github Issues:**

- #302 is the v8 engine crash ticket in this repo whose investigation led to identifying this optimization. #proposeClose
- https://git.yale.edu/lux-its/ml-cluster-formation/issues/52 is another ticket for the same v8 engine crash, but has started to capture interactions with ML Support (https://help.marklogic.com/Tickets/Ticket/View/37229).

**Related links:**

- MarkLogic Support ticket no. 37229

**Wireframe/Mockup:**

N/A

**Comment by brent-hartwig** [comment]
@roamye, I believe this ticket is ready for prioritization.  I associated it to the 9 Sep milestone and provided a default assignee.

**Comment by roamye** [comment]
@brent-hartwig - thanks Brent! I moved it to the 09-23 milestone as the next UAT 09/12 will happen after the 09-09 milestone is deployed.

**Comment by roamye** [comment]
from team meeting [10/25](https://docs.google.com/document/d/1yISUgbiDWvyTcWroxv13pTyM7heSnBprkyAqdVfINso/edit?tab=t.0): 
- A performance test is not needed for this issue as it can be validated through the query console.

**Comment by clarkepeterf** [comment]
Implemented. This is a performance improvement but does not make the original query possible.

The issue with the initial query is with the latter part of the query `{"containingItem": { "memberOf": { "name": "peabody"}}}`. Namely, there are MANY items that match `{ "memberOf": { "name": "peabody"}}`, which causes a failure when those items are used to get sets using the `containingItem` search term.

**Comment by brent-hartwig** [comment]
> This is a performance improvement but does not make the original query possible.

@clarkepeterf, do you mean the search times out?  I just tried in DEV and received results after 42 seconds.  39 seconds the secont time.  I'm not sure why the 20 second timeout did not apply either time.

I could see generating the pre- and post-optimization queries and comparing the performance of each within QC.  I can do this as part of the optimization.  But having identified this optimization while investigating the given query is not one in the same as this optimization having much of an impact on it.

Did you come up with a query or two that helps quantify this optimization?

**Comment by clarkepeterf** [comment]
@brent-hartwig correct - it is still a valid optimization, I just wanted to be clear that the original query that led to its discovery still times out, at least in my experience.

I have been testing using the following:
q: `{"used": {"containingItem": { "name": "blue"}}}`
scope: `event`

**Comment by brent-hartwig** [comment]
First comparison, using the original query.  Probably can ignore the unfiltered runs given these are just for the first 20 results.

```
{"_scope":"event","used":{"containingItem":{"memberOf":{"name":"peabody"}}}}
11.3.1-after-304-filtered firstRun=20313 warmRuns=10 warmMin=17125 warmMax=21694 warmAvg=19361 stddev=2129 totalItemsRead=220
11.3.1-after-304-unfiltered firstRun=19644 warmRuns=10 warmMin=16695 warmMax=21119 warmAvg=18839 stddev=2087 totalItemsRead=220
11.3.1-before-304-filtered firstRun=79529 warmRuns=10 warmMin=25095 warmMax=29855 warmAvg=27003 stddev=2181 totalItemsRead=220
11.3.1-before-304-unfiltered firstRun=78989 warmRuns=10 warmMin=24982 warmMax=29993 warmAvg=26937 stddev=2205 totalItemsRead=220
```

**Comment by roamye** [comment]
Approved by UAT

**Comment by roamye** [comment]
@clarkepeterf - want to confirm this is fine to close.

**Comment by roamye** [comment]
Closing as this is unreviewable.
